File: appMain/src/main/java/com/cartlc/tracker/data/PrefHelper.java
package com.cartlc.tracker.data;

import android.content.Context;
import android.os.Environment;
import android.text.TextUtils;

import com.cartlc.tracker.app.TBApplication;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import timber.log.Timber;

/**
 * Created by dug on 4/17/17.
 */

public class PrefHelper extends PrefHelperBase {

    static PrefHelper sInstance;

    public static PrefHelper getInstance() {
        return sInstance;
    }

    public static void Init(Context ctx) {
        new PrefHelper(ctx);
    }

    static public final String KEY_PROJECT                      = "project";
    static public final String KEY_COMPANY                      = "company";
    static public final String KEY_STREET                       = "street";
    static public final String KEY_STATE                        = "state";
    static public final String KEY_CITY                         = "city";
    static public final String KEY_ZIPCODE                      = "zipcode";
    static final        String KEY_CURRENT_PROJECT_GROUP_ID     = "current_project_group_id";
    static final        String KEY_SAVED_PROJECT_GROUP_ID       = "saved_project_group_id";
    static final        String KEY_FIRST_NAME                   = "first_name";
    static final        String KEY_LAST_NAME                    = "last_name";
    static final        String KEY_TRUCK_NUMBER                 = "truck_number";
    static final        String KEY_LICENSE_PLATE                = "license_plate";
    static final        String KEY_NEXT_PICTURE_COLLECTION_ID   = "next_picture_collection_id";
    static final        String KEY_NEXT_EQUIPMENT_COLLECTION_ID = "next_equipment_collection_id";
    static final        String KEY_NEXT_NOTE_COLLECTION_ID      = "next_note_collection_id";
    static final        String KEY_TECH_ID                      = "tech_id";
    static final        String KEY_REGISTRATION_CHANGED         = "registration_changed";
    static final        String KEY_IS_DEVELOPMENT               = "is_development";

    public static final String VERSION_PROJECT   = "version_project";
    public static final String VERSION_COMPANY   = "version_company";
    public static final String VERSION_EQUIPMENT = "version_equipment";
    public static final String VERSION_NOTE      = "version_note";

    static final String PICTURE_DATE_FORMAT = "yy-MM-dd_HH:mm:ss";
    static final int    VERSION_RESET       = -1;

    PrefHelper(Context ctx) {
        super(ctx);
        sInstance = this;
    }

    public int getTechID() {
        return getInt(KEY_TECH_ID, 0);
    }

    public void setTechID(int id) {
        setInt(KEY_TECH_ID, id);
    }

    public String getStreet() {
        return getString(KEY_STREET, null);
    }

    public void setStreet(String value) {
        setString(KEY_STREET, value);
    }

    public String getState() {
        return getString(KEY_STATE, null);
    }

    public void setState(String value) {
        setString(KEY_STATE, value);
    }

    public String getCompany() {
        return getString(KEY_COMPANY, null);
    }

    public void setCompany(String value) {
        setString(KEY_COMPANY, value);
    }

    public String getCity() {
        return getString(KEY_CITY, null);
    }

    public void setCity(String value) {
        setString(KEY_CITY, value);
    }

    public String getZipCode() {
        return getString(KEY_ZIPCODE, null);
    }

    public void setZipCode(String value) {
        setString(KEY_ZIPCODE, value);
    }

    public String getProjectName() {
        return getString(KEY_PROJECT, null);
    }

    public Long getProjectId() {
        long projectNameId = TableProjects.getInstance().queryProjectName(getProjectName());
        if (projectNameId >= 0) {
            return projectNameId;
        }
        return null;
    }

    public void setProject(String value) {
        setString(KEY_PROJECT, value);
    }

    public long getCurrentProjectGroupId() {
        return getLong(KEY_CURRENT_PROJECT_GROUP_ID, -1L);
    }

    public void setCurrentProjectGroupId(long id) {
        setLong(KEY_CURRENT_PROJECT_GROUP_ID, id);
    }

    public long getSavedProjectGroupId() {
        return getLong(KEY_SAVED_PROJECT_GROUP_ID, -1L);
    }

    public void setSavedProjectGroupId(long id) {
        setLong(KEY_SAVED_PROJECT_GROUP_ID, id);
    }

    public void setFirstName(String name) {
        setString(KEY_FIRST_NAME, name);
    }

    public String getFirstName() {
        return getString(KEY_FIRST_NAME, null);
    }

    public void setLastName(String name) {
        setString(KEY_LAST_NAME, name);
    }

    public String getLastName() {
        return getString(KEY_LAST_NAME, null);
    }

    public boolean hasName() {
        return !TextUtils.isEmpty(getFirstName()) && !TextUtils.isEmpty(getLastName());
    }

    public long getTruckNumber() {
        return getLong(KEY_TRUCK_NUMBER, 0);
    }

    public void setTruckNumber(long id) {
        setLong(KEY_TRUCK_NUMBER, id);
    }

    public String getLicensePlate() {
        return getString(KEY_LICENSE_PLATE, null);
    }

    public void setLicensePlate(String id) {
        setString(KEY_LICENSE_PLATE, id);
    }

    public int getVersionProject() {
        return getInt(VERSION_PROJECT, VERSION_RESET);
    }

    public void setVersionProject(int value) {
        setInt(VERSION_PROJECT, value);
    }

    public int getVersionEquipment() {
        return getInt(VERSION_EQUIPMENT, VERSION_RESET);
    }

    public void setVersionEquipment(int value) {
        setInt(VERSION_EQUIPMENT, value);
    }

    public int getVersionNote() {
        return getInt(VERSION_NOTE, VERSION_RESET);
    }

    public void setVersionNote(int value) {
        setInt(VERSION_NOTE, value);
    }

    public int getVersionCompany() {
        return getInt(VERSION_COMPANY, VERSION_RESET);
    }

    public void setVersionCompany(int value) {
        setInt(VERSION_COMPANY, value);
    }

    public void reloadFromServer() {
        setVersionEquipment(VERSION_RESET);
        setVersionProject(VERSION_RESET);
        setVersionNote(VERSION_RESET);
        setVersionCompany(VERSION_RESET);
    }

    public boolean hasRegistrationChanged() {
        return getInt(KEY_REGISTRATION_CHANGED, 0) != 0;
    }

    public void setRegistrationChanged(boolean flag) {
        setInt(KEY_REGISTRATION_CHANGED, flag ? 1 : 0);
    }

    public boolean isDevelopment() {
        return getInt(KEY_IS_DEVELOPMENT, TBApplication.IsDevelopmentServer() ? 1 : 0) != 0;
    }

    // Return true if added.
    public List<String> addIfNotFound(List<String> list, String element) {
        if (element != null && !list.contains(element)) {
            list.add(element);
            Collections.sort(list);
        }
        return list;
    }

    public DataProjectAddressCombo getCurrentProjectGroup() {
        long projectGroupId = getCurrentProjectGroupId();
        return TableProjectAddressCombo.getInstance().query(projectGroupId);
    }

    public void setupFromCurrentProjectId() {
        DataProjectAddressCombo projectGroup = getCurrentProjectGroup();
        if (projectGroup != null) {
            setProject(projectGroup.getProjectName());
            DataAddress address = projectGroup.getAddress();
            if (address != null) {
                setState(address.state);
                setCity(address.city);
                setCompany(address.company);
                setStreet(address.street);
            }
        }
    }

    public void recoverProject() {
        setCurrentProjectGroupId(getSavedProjectGroupId());
        setupFromCurrentProjectId();
    }

    public void clearCurProject() {
        clearLastEntry();
        setState(null);
        setCity(null);
        setCompany(null);
        setStreet(null);
        setProject(null);
        setSavedProjectGroupId(getCurrentProjectGroupId());
        setCurrentProjectGroupId(-1L);
        TableEquipment.getInstance().clearChecked();
    }

    public void clearLastEntry() {
        setTruckNumber(0);
        setLicensePlate(null);
        TablePictureCollection.getInstance().clearPendingPictures();
        TableNote.getInstance().clearValues();
        TableEquipment.getInstance().clearChecked();
    }

    public boolean hasCurProject() {
        long projectGroupId = getCurrentProjectGroupId();
        if (projectGroupId < 0) {
            return false;
        }
        DataProjectAddressCombo projectGroup = TableProjectAddressCombo.getInstance().query(projectGroupId);
        if (projectGroup == null) {
            return false;
        }
        return true;
    }

    public boolean saveProjectAndAddressCombo() {
        String project = getProjectName();
        if (TextUtils.isEmpty((project))) {
            Timber.i("saveProjectAndAddressCombo(): quit on empty project");
            return false;
        }
        String company = getCompany();
        if (TextUtils.isEmpty(company)) {
            Timber.i("saveProjectAndAddressCombo(): quit on empty company");
            return false;
        }
        String state = getState();
        String street = getStreet();
        String city = getCity();
        String zipcode = getZipCode();
        long addressId;
        addressId = TableAddress.getInstance().queryAddressId(company, street, city, state, zipcode);
        if (addressId < 0) {
            DataAddress address = new DataAddress(company, street, city, state, zipcode);
            address.isLocal = true;
            addressId = TableAddress.getInstance().add(address);
            if (addressId < 0) {
                Timber.i("saveProjectAndAddressCombo(): could not find address: " + address.toString());
            }
        }
        long projectNameId = TableProjects.getInstance().queryProjectName(project);
        if (addressId >= 0 && projectNameId >= 0) {
            long projectGroupId = TableProjectAddressCombo.getInstance().queryProjectGroupId(projectNameId, addressId);
            if (projectGroupId < 0) {
                projectGroupId = TableProjectAddressCombo.getInstance().add(new DataProjectAddressCombo(projectNameId, addressId));
            } else {
                TableProjectAddressCombo.getInstance().updateUsed(projectGroupId);
            }
            setCurrentProjectGroupId(projectGroupId);
            return true;
        }
        Timber.i("saveProjectAndAddressCombo(): could not find project: " + project);
        return false;
    }

    public void setCurrentProjectGroup(DataProjectAddressCombo group) {
        setCurrentProjectGroupId(group.id);
        setProject(group.getProjectName());
        setAddress(group.getAddress());
    }

    void setAddress(DataAddress address) {
        setCompany(address.company);
        setStreet(address.street);
        setCity(address.city);
        setState(address.state);
    }

    public long getNextPictureCollectionID() {
        // Note: ID zero has a special meaning, it means that the set is pending.
        return getLong(KEY_NEXT_PICTURE_COLLECTION_ID, 1L);
    }

    public void incNextPictureCollectionID() {
        setLong(KEY_NEXT_PICTURE_COLLECTION_ID, getNextPictureCollectionID() + 1);
    }

    public long getNextEquipmentCollectionID() {
        return getLong(KEY_NEXT_EQUIPMENT_COLLECTION_ID, 0L);
    }

    public void incNextEquipmentCollectionID() {
        setLong(KEY_NEXT_EQUIPMENT_COLLECTION_ID, getNextEquipmentCollectionID() + 1);
    }

    public long getNextNoteCollectionID() {
        return getLong(KEY_NEXT_NOTE_COLLECTION_ID, 0L);
    }

    public void incNextNoteCollectionID() {
        setLong(KEY_NEXT_NOTE_COLLECTION_ID, getNextEquipmentCollectionID() + 1);
    }

    public DataEntry createEntry() {
        long projectGroupId = getCurrentProjectGroupId();
        if (projectGroupId < 0) {
            return null;
        }
        DataProjectAddressCombo projectGroup = TableProjectAddressCombo.getInstance().query(projectGroupId);
        if (projectGroup == null) {
            return null;
        }
        DataEntry entry = new DataEntry();
        entry.projectAddressCombo = projectGroup;
        entry.equipmentCollection = new DataCollectionEquipmentEntry(getNextEquipmentCollectionID());
        entry.equipmentCollection.addChecked();
        entry.pictureCollection = TablePictureCollection.getInstance().createCollectionFromPending();
        entry.truckNumber = getTruckNumber();
        entry.licensePlateNumber = getLicensePlate();
        entry.saveNotes(getNextNoteCollectionID());
        entry.date = System.currentTimeMillis();
        return entry;
    }

    public String genPictureFilename() {
        long tech_id = getTechID();
        long project_id = getProjectId();
        StringBuilder sbuf = new StringBuilder();
        sbuf.append("picture_t");
        sbuf.append(tech_id);
        sbuf.append("_p");
        sbuf.append(project_id);
        sbuf.append("_d");
        SimpleDateFormat fmt = new SimpleDateFormat(PICTURE_DATE_FORMAT);
        sbuf.append(fmt.format(new Date(System.currentTimeMillis())));
        sbuf.append(".jpg");
        return sbuf.toString();
    }

    public File genFullPictureFile() {
        return new File(mCtx.getExternalFilesDir(Environment.DIRECTORY_PICTURES), genPictureFilename());
    }

    public void clearUploaded() {
        setVersionNote(VERSION_RESET);
        setVersionProject(VERSION_RESET);
        setVersionEquipment(VERSION_RESET);
        setVersionEquipment(VERSION_RESET);
    }

    public String getAddress() {
        StringBuilder sbuf = new StringBuilder();
        String company = getCompany();
        if (company != null) {
            sbuf.append(company);
        }
        String city = getCity();
        if (city != null) {
            if (sbuf.length() > 0) {
                sbuf.append("\n");
            }
            sbuf.append(city);
        }
        String state = getState();
        if (state != null) {
            if (sbuf.length() > 0) {
                sbuf.append(", ");
            }
            sbuf.append(state);
        }
        String zip = getZipCode();
        if (zip != null) {
            if (sbuf.length() > 0) {
                sbuf.append(" ");
            }
            sbuf.append(zip);
        }
        return sbuf.toString();
    }
}

File: appMain/src/main/java/com/cartlc/tracker/model/pref/PrefHelper.kt
/**
 * Copyright 2018, FleetTLC. All rights reserved
 */
package com.cartlc.tracker.model.pref

import android.content.Context
import android.os.Environment
import android.text.TextUtils
import com.cartlc.tracker.model.data.*
import com.cartlc.tracker.model.misc.TruckStatus
import com.cartlc.tracker.model.table.DatabaseTable
import com.cartlc.tracker.ui.app.TBApplication

import java.io.File
import java.text.SimpleDateFormat

import timber.log.Timber
import java.util.*

/**
 * Created by dug on 4/17/17.
 */

class PrefHelper constructor(
        ctx: Context,
        private val db: DatabaseTable
) : PrefHelperBase(ctx) {

    companion object {

        const val KEY_PROJECT = "project"
        const val KEY_COMPANY = "company"
        const val KEY_STREET = "street"
        const val KEY_STATE = "state"
        const val KEY_CITY = "city"
        const val KEY_ZIPCODE = "zipcode"
        const val KEY_TRUCK = "truck" // Number & License
        const val KEY_STATUS = "status"

        private const val KEY_CURRENT_PROJECT_GROUP_ID = "current_project_group_id"
        private const val KEY_SAVED_PROJECT_GROUP_ID = "saved_project_group_id"
        private const val KEY_FIRST_NAME = "first_name"
        private const val KEY_LAST_NAME = "last_name"
        private const val KEY_SECONDARY_FIRST_NAME = "secondary_first_name"
        private const val KEY_SECONDARY_LAST_NAME = "secondary_last_name"
        private const val KEY_HAS_SECONDARY = "has_secondary"
        private const val KEY_TRUCK_NUMBER = "truck_number_string"
        private const val KEY_LICENSE_PLATE = "license_plate"
        private const val KEY_EDIT_ENTRY_ID = "edit_id"
        private const val KEY_NEXT_PICTURE_COLLECTION_ID = "next_picture_collection_id"
        private const val KEY_NEXT_EQUIPMENT_COLLECTION_ID = "next_equipment_collection_id"
        private const val KEY_NEXT_NOTE_COLLECTION_ID = "next_note_collection_id"
        private const val KEY_CURRENT_PICTURE_COLLECTION_ID = "picture_collection_id"
        private const val KEY_TECH_ID = "tech_id"
        private const val KEY_SECONDARY_TECH_ID = "secondary_tech_id"
        private const val KEY_REGISTRATION_CHANGED = "registration_changed"
        private const val KEY_IS_DEVELOPMENT = "is_development"
        private const val KEY_SPECIAL_UPDATE_CHECK = "special_update_check"
        private const val KEY_DO_ERROR_CHECK = "do_error_check"
        private const val KEY_AUTO_ROTATE_PICTURE = "auto_rotate_picture"

        const val VERSION_PROJECT = "version_project"
        const val VERSION_COMPANY = "version_company"
        const val VERSION_EQUIPMENT = "version_equipment"
        const val VERSION_NOTE = "version_note"
        const val VERSION_TRUCK = "version_truck"
        const val VERSION_VEHICLE_NAMES = "version_vehicle_names"

        private const val PICTURE_DATE_FORMAT = "yy-MM-dd_HH:mm:ss"
        private const val VERSION_RESET = -1
    }

    var techID: Int
        get() = getInt(KEY_TECH_ID, 0)
        set(id) = setInt(KEY_TECH_ID, id)

    var secondaryTechID: Int
        get() = getInt(KEY_SECONDARY_TECH_ID, 0)
        set(id) = setInt(KEY_SECONDARY_TECH_ID, id)

    var street: String?
        get() = getString(KEY_STREET, null)
        set(value) = setString(KEY_STREET, value)

    var state: String?
        get() = getString(KEY_STATE, null)
        set(value) = setString(KEY_STATE, value)

    var company: String?
        get() = getString(KEY_COMPANY, null)
        set(value) = setString(KEY_COMPANY, value)

    var city: String?
        get() = getString(KEY_CITY, null)
        set(value) = setString(KEY_CITY, value)

    var zipCode: String?
        get() = getString(KEY_ZIPCODE, null)
        set(value) = setString(KEY_ZIPCODE, value)

    var projectName: String?
        get() = getString(KEY_PROJECT, null)
        set(value) = setString(KEY_PROJECT, value)

    val projectId: Long?
        get() {
            projectName?.let {
                val projectNameId = db.tableProjects.queryProjectName(it)
                return if (projectNameId >= 0) {
                    projectNameId
                } else null
            } ?: return null
        }

    var currentProjectGroupId: Long
        get() = getLong(KEY_CURRENT_PROJECT_GROUP_ID, -1L)
        set(id) = setLong(KEY_CURRENT_PROJECT_GROUP_ID, id)

    var savedProjectGroupId: Long
        get() = getLong(KEY_SAVED_PROJECT_GROUP_ID, -1L)
        set(id) = setLong(KEY_SAVED_PROJECT_GROUP_ID, id)

    var currentPictureCollectionId: Long
        get() = getLong(KEY_CURRENT_PICTURE_COLLECTION_ID, 0)
        set(id) = setLong(KEY_CURRENT_PICTURE_COLLECTION_ID, id)

    var firstName: String?
        get() = getString(KEY_FIRST_NAME, null)
        set(name) = setString(KEY_FIRST_NAME, name)

    var lastName: String?
        get() = getString(KEY_LAST_NAME, null)
        set(name) = setString(KEY_LAST_NAME, name)

    var secondaryFirstName: String?
        get() = getString(KEY_SECONDARY_FIRST_NAME, null)
        set(name) = setString(KEY_SECONDARY_FIRST_NAME, name)

    var secondaryLastName: String?
        get() = getString(KEY_SECONDARY_LAST_NAME, null)
        set(name) = setString(KEY_SECONDARY_LAST_NAME, name)

    var isSecondaryEnabled: Boolean
        get() = getInt(KEY_HAS_SECONDARY, 0) != 0
        set(flag) = setInt(KEY_HAS_SECONDARY, if (flag) 1 else 0)

    var truckNumber: String?
        get() = getString(KEY_TRUCK_NUMBER, null)
        set(id) = setString(KEY_TRUCK_NUMBER, id)

    var licensePlate: String?
        get() = getString(KEY_LICENSE_PLATE, null)
        set(id) = setString(KEY_LICENSE_PLATE, id)

    var versionProject: Int
        get() = getInt(VERSION_PROJECT, VERSION_RESET)
        set(value) = setInt(VERSION_PROJECT, value)

    var versionEquipment: Int
        get() = getInt(VERSION_EQUIPMENT, VERSION_RESET)
        set(value) = setInt(VERSION_EQUIPMENT, value)

    var versionNote: Int
        get() = getInt(VERSION_NOTE, VERSION_RESET)
        set(value) = setInt(VERSION_NOTE, value)

    var versionCompany: Int
        get() = getInt(VERSION_COMPANY, VERSION_RESET)
        set(value) = setInt(VERSION_COMPANY, value)

    var versionTruck: Int
        get() = getInt(VERSION_TRUCK, VERSION_RESET)
        set(value) = setInt(VERSION_TRUCK, value)

    var versionVehicleNames: Int
        get() = getInt(VERSION_VEHICLE_NAMES, VERSION_RESET)
        set(value) = setInt(VERSION_VEHICLE_NAMES, value)

    var status: TruckStatus?
        get() = TruckStatus.from(getInt(KEY_STATUS, TruckStatus.UNKNOWN.ordinal))
        set(status) = if (status == null) {
            setInt(KEY_STATUS, TruckStatus.UNKNOWN.ordinal)
        } else {
            setInt(KEY_STATUS, status.ordinal)
        }

    var currentEditEntryId: Long
        get() = getLong(KEY_EDIT_ENTRY_ID, 0)
        set(id) = setLong(KEY_EDIT_ENTRY_ID, id)

    val currentEditEntry: DataEntry?
        get() = db.tableEntry.query(currentEditEntryId)

    var doErrorCheck: Boolean
        get() = getInt(KEY_DO_ERROR_CHECK, 1) != 0
        set(flag) = setInt(KEY_DO_ERROR_CHECK, if (flag) 1 else 0)

    val isDevelopment: Boolean
        get() = getInt(KEY_IS_DEVELOPMENT, if (TBApplication.IsDevelopmentServer()) 1 else 0) != 0

    var currentProjectGroup: DataProjectAddressCombo?
        get() = db.tableProjectAddressCombo.query(currentProjectGroupId)
        set(group) {
            group?.let {
                currentProjectGroupId = group.id
                projectName = group.projectName
                setAddress(group.address)
            }
        }

    // Note: ID zero has a special meaning, it means that the set is pending.
    val nextPictureCollectionID: Long
        get() = getLong(KEY_NEXT_PICTURE_COLLECTION_ID, 1L)

    val nextEquipmentCollectionID: Long
        get() = getLong(KEY_NEXT_EQUIPMENT_COLLECTION_ID, 0L)

    val nextNoteCollectionID: Long
        get() = getLong(KEY_NEXT_NOTE_COLLECTION_ID, 1L)

    val address: String
        get() {
            val sbuf = StringBuilder()
            val company = company
            if (company != null) {
                sbuf.append(company)
            }
            val street = street
            if (street != null) {
                if (sbuf.length > 0) {
                    sbuf.append("\n")
                }
                sbuf.append(street)
            }
            val city = city
            if (city != null) {
                if (sbuf.length > 0) {
                    sbuf.append("\n")
                }
                sbuf.append(city)
            }
            val state = state
            if (state != null) {
                if (sbuf.length > 0) {
                    sbuf.append(", ")
                }
                sbuf.append(state)
            }
            val zip = zipCode
            if (zip != null) {
                if (sbuf.length > 0) {
                    sbuf.append(" ")
                }
                sbuf.append(zip)
            }
            return sbuf.toString()
        }

    val truckValue: String
        get() {
            val value: String
            val license = licensePlate
            val number = truckNumber
            if (TextUtils.isEmpty(license)) {
                if (number != null) {
                    value = number
                } else {
                    value = ""
                }
            } else {
                if (number != null) {
                    value = DataTruck.toString(number, license)
                } else {
                    value = license ?: ""
                }
            }
            return value
        }

    val numPicturesTaken: Int
        get() = db.tablePictureCollection.countPictures(currentPictureCollectionId)

    val numEquipPossible: Int
        get() {
            val collection = db.tableCollectionEquipmentProject.queryForProject(currentProjectGroup!!.projectNameId)
            return collection.equipment.size
        }

    val autoRotatePicture: Int
        get() = getInt(KEY_AUTO_ROTATE_PICTURE, 0)

    var registrationHasChanged: Boolean
        get() = getInt(KEY_REGISTRATION_CHANGED, 0) != 0
        set(value) {
            setInt(KEY_REGISTRATION_CHANGED, if (value) 1 else 0)
        }

    val hasName: Boolean
        get() = !TextUtils.isEmpty(firstName) && !TextUtils.isEmpty(lastName)

    val hasSecondaryName: Boolean
        get() = !TextUtils.isEmpty(secondaryFirstName) && !TextUtils.isEmpty(secondaryLastName)

    fun getKeyValue(key: String): String? {
        return getString(key, null)
    }

    fun setKeyValue(key: String, value: String?) {
        setString(key, value)
    }

    fun reloadFromServer() {
        versionEquipment = VERSION_RESET
        versionProject = VERSION_RESET
        versionNote = VERSION_RESET
        versionCompany = VERSION_RESET
        versionTruck = VERSION_RESET
    }

    fun detectSpecialUpdateCheck() {
        val value = getInt(KEY_SPECIAL_UPDATE_CHECK, 0)
        if (value < 1) {
            reloadFromServer()
            setInt(KEY_SPECIAL_UPDATE_CHECK, 1)
        }
    }


    fun setFromCurrentProjectId() {
        val projectGroup = currentProjectGroup
        if (projectGroup != null) {
            projectName = projectGroup.projectName
            val address = projectGroup.address
            if (address != null) {
                setAddress(address)
            }
        }
    }

    fun clearCurProject() {
        clearLastEntry()
        state = null
        city = null
        company = null
        street = null
        zipCode = null
        projectName = null
        savedProjectGroupId = currentProjectGroupId
        currentProjectGroupId = -1L
        db.tableEquipment.clearChecked()
    }

    fun clearLastEntry() {

        db.reportDebugMessage("clearLastEntry(): TRUCK#=$truckNumber, #PICTURES="
                + db.tablePictureCollection.countPendingPictures())

        truckNumber = null
        licensePlate = null
        setKeyValue(KEY_TRUCK, null)
        status = null
        currentEditEntryId = 0
        currentPictureCollectionId = 0
        db.tablePictureCollection.clearPendingPictures()
        db.tableNote.clearValues()
        db.tableEquipment.clearChecked()
    }

    fun saveProjectAndAddressCombo(modifyCurrent: Boolean): Boolean {
        val project = projectName
        if (project.isNullOrBlank()) {
            Timber.e("project name cannot be blank")
            return false
        }
        val company = company
        if (company.isNullOrBlank()) {
            Timber.e("company cannot be blank")
            return false
        }
        val state = state
        if (state.isNullOrBlank()) {
            Timber.e("state cannot be blank.")
            return false
        }
        val street = street
        if (street.isNullOrBlank()) {
            Timber.e("street cannot be blank.")
            return false
        }
        val city = city
        if (city.isNullOrBlank()) {
            Timber.e("city cannot be blank.")
            return false
        }
        val zipcode = zipCode
        var addressId: Long
        addressId = db.tableAddress.queryAddressId(company, street, city, state, zipcode)
        if (addressId < 0) {
            val address = DataAddress(company, street, city, state, zipcode)
            address.isLocal = true
            addressId = db.tableAddress.add(address)
            if (addressId < 0) {
                Timber.e("saveProjectAndAddressCombo(): could not find address: " + address.toString())
                clearCurProject()
                return false
            }
        }
        val projectNameId = db.tableProjects.queryProjectName(project)
        if (projectNameId < 0) {
            Timber.e("saveProjectAndAddressCombo(): could not find project: $project")
            clearCurProject()
            return false
        }
        var projectGroupId: Long
        if (modifyCurrent) {
            projectGroupId = currentProjectGroupId
            if (projectGroupId < 0) {
                Timber.e("saveProjectAndAddressCombo(): could not modify current project, none is alive")
                return false
            }
            val combo = db.tableProjectAddressCombo.query(projectGroupId) ?: return false
            combo.reset(projectNameId, addressId)
            if (!db.tableProjectAddressCombo.save(combo)) {
                Timber.e("saveProjectAndAddressCombo(): could not update project combo")
                return false
            }
            db.tableProjectAddressCombo.mergeIdenticals(combo)
            val count = db.tableEntry.reUploadEntries(combo)
            if (count > 0) {
                Timber.i("saveProjectAddressCombo(): re-upload $count entries")
            }
            db.tableProjectAddressCombo.updateUsed(projectGroupId)
        } else {
            projectGroupId = db.tableProjectAddressCombo.queryProjectGroupId(projectNameId, addressId)
            if (projectGroupId < 0) {
                projectGroupId = db.tableProjectAddressCombo.add(DataProjectAddressCombo(db, projectNameId, addressId))
            } else {
                db.tableProjectAddressCombo.updateUsed(projectGroupId)
            }
            currentProjectGroupId = projectGroupId
        }
        return true
    }

    private fun setAddress(address: DataAddress?) {
        company = address!!.company
        street = address.street
        city = address.city
        state = address.state
        zipCode = address.zipcode
    }

    fun incNextPictureCollectionID() {
        setLong(KEY_NEXT_PICTURE_COLLECTION_ID, nextPictureCollectionID + 1)
    }

    fun incNextEquipmentCollectionID() {
        setLong(KEY_NEXT_EQUIPMENT_COLLECTION_ID, nextEquipmentCollectionID + 1)
    }

    fun incNextNoteCollectionID() {
        setLong(KEY_NEXT_NOTE_COLLECTION_ID, nextEquipmentCollectionID + 1)
    }

    fun createEntry(): DataEntry? {
        val projectGroupId = currentProjectGroupId
        if (projectGroupId < 0) {
            return null
        }
        val projectGroup = db.tableProjectAddressCombo.query(projectGroupId) ?: return null
        val entry = DataEntry(db)
        entry.projectAddressCombo = projectGroup
        entry.equipmentCollection = DataCollectionEquipmentEntry(db, nextEquipmentCollectionID)
        entry.equipmentCollection!!.addChecked()
        entry.pictureCollection = db.tablePictureCollection.createCollectionFromPending(nextPictureCollectionID)
        entry.truckId = db.tableTruck.save(
                truckNumber ?: "",
                licensePlate ?: "",
                projectGroup.projectNameId,
                projectGroup.companyName!!)
        entry.status = status
        entry.saveNotes(nextNoteCollectionID)
        entry.date = System.currentTimeMillis()
        return entry
    }

    fun setFromEntry(entry: DataEntry) {
        currentEditEntryId = entry.id
        currentProjectGroupId = entry.projectAddressCombo!!.id
        currentPictureCollectionId = entry.pictureCollection!!.id
        entry.equipmentCollection!!.setChecked()
        val truck = entry.truck
        if (truck != null) {
            truckNumber = truck.truckNumber
            licensePlate = truck.licensePlateNumber
        } else {
            truckNumber = null
            licensePlate = null
        }
        status = entry.status
        db.tableNote.clearValues()
    }

    fun saveEntry(): DataEntry? {
        val entry = currentEditEntry ?: return createEntry()
        entry.equipmentCollection!!.addChecked()
        var truck = entry.truck
        if (truck == null) {
            truck = DataTruck()
        }
        truck.truckNumber = truckNumber
        truck.licensePlateNumber = licensePlate
        truck.hasEntry = true
        entry.truckId = db.tableTruck.save(truck)
        entry.status = status
        entry.uploadedMaster = false
        entry.uploadedAws = false
        entry.hasError = false
        entry.serverErrorCount = 0.toShort()
        // Be careful here: I use the date to match an tableEntry when looking up the server id for older APP versions.
        if (entry.serverId > 0) {
            entry.date = System.currentTimeMillis()
        }
        entry.saveNotes()
        return entry
    }

    fun genPictureFilename(): String {
        val tech_id = techID.toLong()
        val project_id = projectId!!
        val sbuf = StringBuilder()
        sbuf.append("picture_t")
        sbuf.append(tech_id)
        sbuf.append("_p")
        sbuf.append(project_id)
        sbuf.append("_d")
        val fmt = SimpleDateFormat(PICTURE_DATE_FORMAT, Locale.getDefault())
        sbuf.append(fmt.format(Date(System.currentTimeMillis())))
        sbuf.append(".jpg")
        return sbuf.toString()
    }

    fun genFullPictureFile(): File {
        return File(ctx.getExternalFilesDir(Environment.DIRECTORY_PICTURES), genPictureFilename())
    }

    fun clearUploaded() {
        versionNote = VERSION_RESET
        versionProject = VERSION_RESET
        versionEquipment = VERSION_RESET
        versionCompany = VERSION_RESET
        versionTruck = VERSION_RESET
    }

    fun reloadProjects() {
        versionProject = VERSION_RESET
    }

    fun reloadEquipments() {
        versionEquipment = VERSION_RESET
    }

    fun parseTruckValue(value: String) {
        if (value.contains(":")) {
            val ele = value.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (ele.size >= 2) {
                truckNumber = ele[0].trim { it <= ' ' }
                licensePlate = ele[1].trim { it <= ' ' }
            } else if (ele.size == 1) {
                if (TextUtils.isDigitsOnly(ele[0])) {
                    truckNumber = ele[0].trim { it <= ' ' }
                } else {
                    licensePlate = ele[0].trim { it <= ' ' }
                }
            } else {
                truckNumber = null
                licensePlate = null
            }
        } else if (TextUtils.isDigitsOnly(value)) {
            truckNumber = value
        } else {
            licensePlate = value
        }
    }

    fun incAutoRotatePicture(degrees: Int) {
        val newValue = (autoRotatePicture + degrees) % 360
        setInt(KEY_AUTO_ROTATE_PICTURE, newValue)
    }

    fun clearAutoRotatePicture() {
        setInt(KEY_AUTO_ROTATE_PICTURE, 0)
    }

}
